elements/dash-tabs/dash-tabs.js

```diff
@@ -0,0 +1,293 @@
+/**
+ * Tab panels are a pattern to limit visible content by separating
+ * it into multiple panels. Only one panel is visible at a time, while
+ * _all_ corresponding tabs are always visible. To switch from one panel
+ * to another, the corresponding tab has to be selected.
+ *
+ * By either clicking or by using the arrow keys the user changes the
+ * selection of the active tab.
+ *
+ * If JavaScript is disabled, all panels are shown interleaved with the
+ * respective tabs. The tabs now function as headings.
+ */
+(function() {
+  /**
+   * Define keycodes to help with handling keyboard events.
+   */
+  const KEYCODE = {
+    DOWN: 40,
+    LEFT: 37,
+    RIGHT: 39,
+    UP: 38,
+    HOME: 36,
+    END: 35,
+  };
+
+  /**
+   * `DashTabs` is a container element for tabs and panels.
+   *
+   * All children of `<dash-tabs>` should be either `<dash-tab>` or
+   * `<dash-tabpanel>`. This element is stateless, meaning that no values are
+   * cached and therefore, changes during runtime work.
+   *
+   * Because both the tabs and the panels link to each other semantically, every
+   * tab and every panel _must_ have a unique ID.
+   */
+  class DashTabs extends HTMLElement {
+    constructor() {
+      super();
+    }
+
+    /**
+     * `connectedCallback` groups tabs and panels by reordering and makes sure
+     * exactly one tab is active.
+     */
+    connectedCallback() {
+      // The element needs to do some manual input event handling to allow
+      // switching with arrow keys and Home/End.
+      this.addEventListener('keydown', this._onKeyDown);
+      this.addEventListener('click', this._onClick);
+
+      this.setAttribute('role', 'tablist');
+      // Acquire all tabs and panels inside the element
+      const tabs = this._allTabs();
+      const panels = this._allPanels();
+      // If there are no tabs, there is no way to switch between panels. Abort.
+      if (!tabs.length) return;
+
+      // For progressive enhancement, the markup should alternate between tabs
+      // and panels. If JavaScript is disabled, all panels are
+      // visible with their respective tab right above them.
+      // If JavaScript is enabled, the element groups all children by type.
+      // First all the tabs, then all the panels.
+      // Calling `appendChild` on an already inserted element _moves_ the
+      // element to the last position.
+      tabs.forEach(tab => this.appendChild(tab));
+      panels.forEach(panel => this.appendChild(panel));
+
+      // Give each panel a `aria-labelledby` attribute that refers to the tab
+      // that controls it.
+      tabs.forEach(tab => {
+        if (tab.id)
```

Comment:
>  If our target audience is component authors, this might be a useful pattern to teach them.

Fair point. I’ll add it.

> I don't know if I'd import a library that generates a random ID for me.

No, I wouldn’t, either. Counter or time-based should be fine. I’ll whip something up :)